### PR TITLE
Add premultiplied bulk pixel operations

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
This adds new methods to `PixelOperations` which allows optimized bulk conversion to and from premultiplied pixels. 

I've managed to merge the generic operations into a single loop but not the Rgba SIMD optimized versions. It would be ace if someone could manage it but I don't think we have the tools as is.

### CODE REVIEW

File: src/ImageSharp/PixelFormats/Utils/Vector4Converters.Default.cs [hunk 1]
```diff
@@ -87,6 +123,42 @@ public static class Default
...
+                where TPixel : struct, IPixel<TPixel>
+            {
+                ref Vector4 sourceRef = ref MemoryMarshal.GetReference(sourceVectors);
+                ref TPixel destRef = ref MemoryMarshal.GetReference(destinationColors);
+
+                for (int i = 0; i < sourceVectors.Length; i++)
+                {
+                    ref Vector4 sp = ref Unsafe.Add(ref sourceRef, i);
+                    ref TPixel dp = ref Unsafe.Add(ref destRef, i);
+                    Vector4Utils.UnPremultiply(ref sp);
```
Comment (antonfirsov) [inline src/ImageSharp/PixelFormats/Utils/Vector4Converters.Default.cs]: Is this mixed loop as default generic implementation faster than a sequence of bulk conversions + premultiplications? Could be only checked by a benchmark.

Comment (JimBobSquarePants) [reply src/ImageSharp/PixelFormats/Utils/Vector4Converters.Default.cs]: Logic would dictate yes but I can add a rough benchmark to test I suppose.

**Comment by JimBobSquarePants** [comment]
@antonfirsov Agree with all the above. 

Forgot to add WIP. I want as many eyes on this as possible as I develop it and hopefully some assistance as I'm off to Seattle on Sunday.

File: src/ImageSharp/PixelFormats/Utils/Vector4Converters.Default.cs (hunk 1 above)
Comment (antonfirsov) [reply src/ImageSharp/PixelFormats/Utils/Vector4Converters.Default.cs]: CPU caches and pipelines follow a different kind of logic sometimes :)

**Comment by JimBobSquarePants** [comment]
@antonfirsov We're doing something weird in `ResizeProcessor` when companding.

Our pattern is. 

`premultiply => expand => operate => unpremultiply => compress`

When it should be (requires change to companding algorithm)
`premultiply => expand => operate => compress => unpremultiply`

or 
`expand => premultiply => operate => unpremultiply => compress`

File: tests/ImageSharp.Tests/PixelFormats/PixelOperationsTests.Blender.cs [hunk 2]
```diff
@@ -1,48 +1,48 @@
...
-using System.Text;
-using SixLabors.ImageSharp.PixelFormats;
-using SixLabors.ImageSharp.PixelFormats.PixelBlenders;
-using SixLabors.ImageSharp.Tests.TestUtilities;
-using Xunit;
-
-namespace SixLabors.ImageSharp.Tests.PixelFormats
-{
-    public class PixelBlenderTests
+﻿// Copyright (c) Six Labors and contributors.
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/PixelFormats/PixelOperationsTests.Blender.cs]: Fixing line endings only.

**Comment by JimBobSquarePants** [comment]
Ready for review. Coverage is good. Codecov just can't compare as the master build failed due to System.Drawing.

**Comment by antonfirsov** [comment]
@JimBobSquarePants gonna review this extensively in the evening or tomorrow. Won't be easy and quick round, I can see many points where perf. could be hurt.

**Comment by JimBobSquarePants** [comment]
@antonfirsov I think you'll be pleasantly surprised actually. Here's an (unplugged laptop cos it's 6am, I'm jetlagged, no-one else is awake, the cable is in a room where others are sleeping) benchmark where I would expect a potential hit (Resize + Companding) 

``` bash
BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17763.379 (1809/October2018Update/Redstone5)
Intel Core i7-6600U CPU 2.60GHz (Skylake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=2.2.202
  [Host] : .NET Core 2.1.9 (CoreCLR 4.6.27414.06, CoreFX 4.6.27415.01), 64bit RyuJIT
  Clr    : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3362.0
  Core   : .NET Core 2.1.9 (CoreCLR 4.6.27414.06, CoreFX 4.6.27415.01), 64bit RyuJIT

IterationCount=3  LaunchCount=1  WarmupCount=3

                                   Method |  Job | Runtime | SourceSize | DestSize |      Mean |      Error |     StdDev | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
----------------------------------------- |----- |-------- |----------- |--------- |----------:|-----------:|-----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
                            SystemDrawing |  Clr |     Clr |       3032 |      400 | 115.72 ms |  19.455 ms |  1.0664 ms |  1.00 |    0.00 |           - |           - |           - |              1638 B |
 'ImageSharp, MaxDegreeOfParallelism = 1' |  Clr |     Clr |       3032 |      400 | 135.83 ms | 355.538 ms | 19.4882 ms |  1.17 |    0.18 |           - |           - |           - |             16384 B |
 'ImageSharp, MaxDegreeOfParallelism = 4' |  Clr |     Clr |       3032 |      400 |  58.82 ms |  10.872 ms |  0.5959 ms |  0.51 |    0.01 |           - |           - |           - |             23666 B |
 'ImageSharp, MaxDegreeOfParallelism = 8' |  Clr |     Clr |       3032 |      400 |  66.90 ms |  82.627 ms |  4.5291 ms |  0.58 |    0.04 |           - |           - |           - |             23666 B |
                                          |      |         |            |          |           |            |            |       |         |             |             |             |                     |
                            SystemDrawing | Core |    Core |       3032 |      400 | 134.72 ms | 127.207 ms |  6.9727 ms |  1.00 |    0.00 |           - |           - |           - |                96 B |
 'ImageSharp, MaxDegreeOfParallelism = 1' | Core |    Core |       3032 |      400 | 122.37 ms |  16.997 ms |  0.9317 ms |  0.91 |    0.05 |           - |           - |           - |             15704 B |
 'ImageSharp, MaxDegreeOfParallelism = 4' | Core |    Core |       3032 |      400 |  55.14 ms |  37.837 ms |  2.0740 ms |  0.41 |    0.01 |           - |           - |           - |             19592 B |
 'ImageSharp, MaxDegreeOfParallelism = 8' | Core |    Core |       3032 |      400 |  55.13 ms |   7.451 ms |  0.4084 ms |  0.41 |    0.02 |           - |           - |           - |             20028 B |
```

File: src/ImageSharp/PixelFormats/Utils/Vector4Converters.Default.cs (hunk 1 above)
Comment (antonfirsov) [reply src/ImageSharp/PixelFormats/Utils/Vector4Converters.Default.cs]: This is most likely OK for the default path, but we should be aware that there is some kind of individual threshold for the number of operations it is OK to put inside a loop without making it slower than doing separate bulk operations instead.

File: src/ImageSharp/PixelFormats/PixelOperations{TPixel}.cs [hunk 3]
```diff
@@ -61,33 +60,169 @@ public partial class PixelOperations<TPixel>
...
+            Utils.Vector4Converters.Default.DangerousFromPremultipliedVector4(sourceVectors, destPixels);
+        }
+
+        /// <summary>
+        /// Bulk version of <see cref="IPixel.ToVector4()"/> converting 'sourceColors.Length' pixels into alpha premultiplied 'destinationVectors'.
+        /// </summary>
+        /// <param name="configuration">A <see cref="Configuration"/> to configure internal operations</param>
+        /// <param name="sourcePixels">The <see cref="Span{T}"/> to the source colors.</param>
+        /// <param name="destVectors">The <see cref="Span{T}"/> to the destination vectors.</param>
+        internal virtual void ToPremultipliedVector4(
```
Comment (antonfirsov) [inline src/ImageSharp/PixelFormats/PixelOperations{TPixel}.cs]: Unlike `ToVector4()` this is only implemented by `Rgba32.PixelOperations`. Currently, using this method in `ResizeProcessor` wires out the optimizations for the remaining "RGBA - like" pixel types (`Argb32`, `Bgra32`, `Rgb24`, `Bgr24`). (They are delegating to `Vector4Converters.RgbaCompatible.ToVector4()`).

File: src/ImageSharp/ColorSpaces/Companding/SRgbCompanding.cs [hunk 4]
```diff
@@ -48,27 +46,23 @@ public static void Compress(Span<Vector4> vectors)
...
         }
 
         /// <summary>
         /// Expands a companded vector to its linear equivalent with respect to the energy.
         /// </summary>
         /// <param name="vector">The vector.</param>
-        /// <returns>The <see cref="Vector4"/> representing the linear channel values.</returns>
         [MethodImpl(InliningOptions.ShortMethod)]
-        public static Vector4 Expand(Vector4 vector) => new Vector4(Expand(vector.X), Expand(vector.Y), Expand(vector.Z), vector.W);
+        public static void Expand(ref Vector4 vector) => vector = new Vector4(Expand(vector.X), Expand(vector.Y), Expand(vector.Z), vector.W);
```
Comment (antonfirsov) [inline src/ImageSharp/ColorSpaces/Companding/SRgbCompanding.cs]: The unnecessary constructor invocation + copy is still there. What I meant is:

```C#
void Expand(ref Vector4 v) 
{
    v.X =Expand(v.X);
    v.Y =Expand(v.Y);
    v.Z =Expand(v.Z);
    v.W =Expand(v.W);    
}
```

It's unlikely that JIT is capable to optimize the simple version. Try a benchmark for a proof/disproof! :smile:

File: src/ImageSharp/ColorSpaces/Companding/SRgbCompanding.cs [hunk 5]
```diff
@@ -60,15 +56,15 @@ public static void Compress(Span<Vector4> vectors)
         /// <param name="vector">The vector.</param>
         /// <returns>The <see cref="Vector4"/> representing the linear channel values.</returns>
         [MethodImpl(InliningOptions.ShortMethod)]
-        public static Vector4 Expand(Vector4 vector) => new Vector4(Expand(vector.X), Expand(vector.Y), Expand(vector.Z), vector.W);
+        public static Vector4 Expand(ref Vector4 vector) => vector = new Vector4(Expand(vector.X), Expand(vector.Y), Expand(vector.Z), vector.W);
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/ColorSpaces/Companding/SRgbCompanding.cs]: At least that's an easy fix.

File: src/ImageSharp/PixelFormats/PixelOperations{TPixel}.cs (hunk 3 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/PixelFormats/PixelOperations{TPixel}.cs]: Ah yes, the fix should be fairly trivial then. We simply need to ensure that all the overloads we use in `Rgba32.PixelOperations` are also overloaded in our `RgbaCompatible` generated code.

File: src/ImageSharp/ColorSpaces/Companding/SRgbCompanding.cs (hunk 4 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/ColorSpaces/Companding/SRgbCompanding.cs]: Yeah, I got it. Already committed a change locally. 👍

**Comment by JimBobSquarePants** [comment]
@antonfirsov Good spot with the API. That's much better now!

``` bash
BenchmarkDotNet=v0.11.3, OS=Windows 10.0.17763.379 (1809/October2018Update/Redstone5)
Intel Core i7-6600U CPU 2.60GHz (Skylake), 1 CPU, 4 logical and 2 physical cores
.NET Core SDK=2.2.202
  [Host] : .NET Core 2.1.9 (CoreCLR 4.6.27414.06, CoreFX 4.6.27415.01), 64bit RyuJIT
  Clr    : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3362.0
  Core   : .NET Core 2.1.9 (CoreCLR 4.6.27414.06, CoreFX 4.6.27415.01), 64bit RyuJIT

IterationCount=3  LaunchCount=1  WarmupCount=3

                                   Method |  Job | Runtime | SourceSize | DestSize |      Mean |     Error |    StdDev | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
----------------------------------------- |----- |-------- |----------- |--------- |----------:|----------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
                            SystemDrawing |  Clr |     Clr |       3032 |      400 | 106.93 ms |  5.611 ms | 0.3076 ms |  1.00 |    0.00 |           - |           - |           - |              1638 B |
 'ImageSharp, MaxDegreeOfParallelism = 1' |  Clr |     Clr |       3032 |      400 | 100.97 ms | 37.292 ms | 2.0441 ms |  0.94 |    0.02 |           - |           - |           - |             16384 B |
                                          |      |         |            |          |           |           |           |       |         |             |             |             |                     |
                            SystemDrawing | Core |    Core |       3032 |      400 | 118.23 ms | 24.147 ms | 1.3236 ms |  1.00 |    0.00 |           - |           - |           - |                96 B |
 'ImageSharp, MaxDegreeOfParallelism = 1' | Core |    Core |       3032 |      400 |  93.31 ms | 12.256 ms | 0.6718 ms |  0.79 |    0.00 |           - |           - |           - |             15704 B |
```

File: tests/ImageSharp.Tests/PixelFormats/PixelOperations/PixelOperationsTests.cs [hunk 6]
```diff
@@ -477,25 +681,37 @@ public void ToRgba64Bytes(int count)
             );
         }
 
-        internal static Vector4[] CreateExpectedVector4Data(TPixel[] source)
+        public delegate void RefAction<T1>(ref T1 arg1);
```
Comment (antonfirsov) [inline tests/ImageSharp.Tests/PixelFormats/PixelOperations/PixelOperationsTests.cs]: Nice! 👍

**Comment by JimBobSquarePants** [comment]
> The From*** overloads are still missing for the premultiplied/compand stuff in RGBA-like types. (FromPremultipliedVector4, FromPremultipliedScaledVector4, FromCompandedScaledVector4, FromCompandedPremultipliedScaledVector4)
I strongly suggest to deal with all this boilerplate in this (or a follow-up) PR to avoid maintenance hell. I think enum flags could do the job without hurting performance. For bulk operations the overhead should be negligable.

I deliberately didn't add those overloads because they expect `ReadOnlySpan<Vector4>` and we have no optimization path in the overloads that allows us intercept the copy and the `Vector4` elements before passing to the `TPixel`. It's all bulk operations.

If you can think of some workaround please push because I couldn't figure one out and chose to wait for us to optimize the default implementations.

**Comment by antonfirsov** [comment]
## 1. `From***Vector4(...)`
This is more a design question than a question of implementation tricks or workarounds. This is mostly a design PR, so we need to sort out design + API issues before merging.

I have proposal!

#### Observation
When an image processing code calls `.From***Vector4(sourceVectors, destPixels )` it is a final phase of the given algorithm and the data in `sourceVectors` is no longer needed. To me it looks like this is an invariant which is true for the 100% of our current processors+encoders+decoders.

#### Conclusion
`.From***Vector4(sourceVectors, destPixels )` could be a destructive operation and modifying the contents of the `sourceVectors` buffer is OK. It could be a `Span<T>` instead of `ReadOnlySpan<T>`. What we need is good naming and documentation, to make this obvious on the call site.

## 2. Eliminating boilerplate
The more I think about this topic the more it feels it should be the part of this design PR, to minimize the amount of technical debt in our codebase. With my flags proposal it's possible to collapse the current amount of `2*2³ = 16` operations (per `PixelOperations<T>` implementation!) into **a pair of `.FromVector4(...)` + `.ToVector4(...)` functions**. What do you think?

I would be happy to help you out with the implementation, but unfortunately it's not possible for me for another 2 weeks because of family happenings and a daily job business trip 😞

**Comment by JimBobSquarePants** [comment]
@antonfirsov I did actually consider making the `From**Vector4` methods destructive having come to the same conclusion as you but it left me with an uneasy feeling to do so. I could yet be convinced as the current format is far too limiting performance-wise. 

I don't know if I have the chops to do the flags based refactoring work on my own (I'd really like to reduce the API surface) so am happy to wait. I have my eye on a couple of other things in the codebase that I can fix up in the interim.

@dlemstra @tocsoft What are your thoughts?

**Comment by JimBobSquarePants** [comment]
Closing in favour of #869

**Pull request closed without merging**